src/stream/src/executor/join/row.rs

```diff
@@ -47,10 +48,16 @@ impl<R: Row> JoinRow<R> {
         (&self.row, degree)
     }
 
-    pub fn encode(&self) -> EncodedJoinRow {
-        EncodedJoinRow {
-            compacted_row: (&self.row).into(),
-            degree: self.degree,
+    pub fn encode<const N: JoinEncodingPrimitive>(&self) -> CachedJoinRow {
```

Comment:
By making `JoinEncodingPrimitive` a const generic we doubled the code to be compiled after monomorphization, but here we fit different variants back to a single `enum` type and use runtime dispatching when calling method on it. This doesn't seem to be a good deal. Could you try either...

- use generic `CachedJoinRow` here (like by associating a type to the `JoinEncoding` parameter)
- or simply use a boolean/enum field to indicate the encoding type?

Reply:
updated, I thought the compiler was going to do some magic about the `const N` that only one code path will be kept for each instantiation of the function with respect to each choice of `N`

Now it is dispatched during compilation, 
But not sure if it is exactly what `use generic CachedJoinRow here (like by associating a type to the JoinEncoding parameter)` suggests

Reply:
Oh I don't mean this. I mean we should avoid introducing an enum to erase the type under different `JoinEncoding`. Instead, we should use different types to avoid any runtime dispatching. The trait can be like this:


```rust
pub trait JoinEncoding: 'static + Send + Sync {
    type EncodedRow;

    fn encode<R: Row>(row: &JoinRow<R>) -> Self::EncodeRow;
}
```

Reply:
updated